### PR TITLE
Added optional regex support

### PR DESCRIPTION
**Pull request by uroybd**
Hi,
I've just added optional regex support which doesn't break anything, **totally compatible** with the previous version but you can add a third optional regex to break by that regex instead of breaking into letters. It is particularly **useful** for Indic languages(e.g. Bengali, Hindi, Tamil) and Arabic.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,27 +1,33 @@
...
-    while (++i < length) {
+    var parentNode = element.parentNode
+    var letter_data = []
+    if (regex) {
+      letter_data = string.match(regex)
+    } else {
+      letter_data = string.split('')
+    }
+    var length = letter_data.length
+    for (var i = 0; i < length; i++) {
```
Comment (yuanqing) [inline index.js]: Revert to the original way of writing a loop:

```js
var length = split.length
var i = -1
while (++i < length) {
  // ...
}
```

File: index.js [hunk 2]
```diff
@@ -1,27 +1,33 @@
...
       if (classPrefix) {
         node.className = classPrefix + count
         count++
       }
-      node.appendChild(document.createTextNode(string[i]))
+      node.appendChild(document.createTextNode(letter_data[i]))
       node.setAttribute('aria-hidden', 'true')
       parentNode.insertBefore(node, element)
     }
-    if (string.trim() !== '') {
```
Comment (yuanqing) [inline index.js]: Revert this change

File: index.js [hunk 3]
```diff
@@ -1,27 +1,33 @@
...
+      letter_data = string.split('')
+    }
+    var length = letter_data.length
+    for (var i = 0; i < length; i++) {
       var node = document.createElement(tagName)
       if (classPrefix) {
         node.className = classPrefix + count
         count++
       }
-      node.appendChild(document.createTextNode(string[i]))
```
Comment (yuanqing) [inline index.js]: Instead of `string[i]`, we would then have `split[i]` here

File: test.js [hunk 4]
```diff
@@ -100,3 +100,18 @@ test('can inject spans with a custom class prefix', function (t) {
...
+    '<span>আদর্শ লিপি</span>'
+  );
+  var bengaliUnitRegex = /([অ-হ](্[অ-হ])+|[অ-হ]্|[অ-হ])[া-ৌ]*ঁ*|ৎ| /g
+  charming(element, {}, bengaliUnitRegex);
+  t.equal(element.getAttribute('aria-label'), null)
+  t.equal(
+    element.innerHTML,
+    '<span aria-label="আদর্শ লিপি"><span class="char1" aria-hidden="true">আ</span><span class="char2" aria-hidden="true">দ</span><span class="char3" aria-hidden="true">র্শ</span><span class="char4" aria-hidden="true"> </span><span class="char5" aria-hidden="true">লি</span><span class="char6" aria-hidden="true">পি</span></span>'
+  );
+});
```
Comment (yuanqing) [inline test.js]: Suggest having a simplified test case with a simple regular expression

File: index.js [hunk 5]
```diff
@@ -1,27 +1,33 @@
...
-    var length = string.length
-    var i = -1
-    while (++i < length) {
+    var parentNode = element.parentNode
+    var letter_data = []
+    if (regex) {
+      letter_data = string.match(regex)
+    } else {
+      letter_data = string.split('')
+    }
```
Comment (yuanqing) [inline index.js]: Prefer using the following for splitting the `string`:

```js
var split = splitRegex ? string.split(splitRegex) : string
```

File: index.js [hunk 6]
```diff
@@ -1,27 +1,33 @@
-module.exports = function (element, options) {
+module.exports = function (element, options, regex) {
```
[the remark this answers is not in the document]
Comment (uroybd) [reply index.js]: That's cool. :)

File: index.js [hunk 7]
```diff
@@ -1,27 +1,33 @@
...
   options = options || {}
   element.normalize()
+  regex = regex || null
```
[the remark this answers is not in the document]
Comment (uroybd) [reply index.js]: Okay.

File: index.js [hunk 8]
```diff
@@ -1,27 +1,33 @@
...
 
   var tagName = options.tagName || 'span'
   var classPrefix = options.classPrefix != null ? options.classPrefix : 'char'
   var count = 1
 
   function inject (element) {
-    var parentNode = element.parentNode
```
[the remark this answers is not in the document]
Comment (uroybd) [reply index.js]: It's already there. Only 1 line below.

File: index.js (hunk 1 above)
Comment (uroybd) [reply index.js]: Okay

File: index.js (hunk 2 above)
Comment (uroybd) [reply index.js]: Okay.

File: index.js (hunk 3 above)
Comment (uroybd) [reply index.js]: Yes.

File: index.js (hunk 5 above)
Comment (uroybd) [reply index.js]: Sure

File: test.js (hunk 4 above)
Comment (uroybd) [reply test.js]: Ah! It's simple, trust me, for Bengali, it's the simplest!

**Comment by uroybd** [comment]
Hello, changed as you proposed except the regex test. It is the simplest for Bengali.

File: test.js (hunk 4 above)
Comment (yuanqing) [reply test.js]: @uroybd 

This test failed https://travis-ci.org/yuanqing/charming/jobs/340147531#L520. (There’s an issue in the Travis script which I will fix separately.)

Am still of the opinion that this regular expression with non-alphabetic characters and character sets is extremely complex, and will be quite difficult to maintain going forward.

This PR actually addresses issue #1; can the test be a super simple one of splitting on words? For example:

```js
const element = createElement('foo bar')
charming(element, {
  splitRegex: /(\s+)/,
  classPrefix: 'word'
})
t.equal(element.getAttribute('aria-label'), 'foo bar')
t.equal(
  element.innerHTML,
  '<span class="word1" aria-hidden="true">foo</span><span class="word2" aria-hidden="true"> </span><span class="word3" aria-hidden="true">bar</span>'
)
```

File: index.js [hunk 9]
```diff
@@ -9,15 +10,16 @@ module.exports = function (element, options) {
   function inject (element) {
     var parentNode = element.parentNode
     var string = element.nodeValue
-    var length = string.length
+    var split = splitRegex ? string.split(splitRegex) : string
+    var length = split.length
     var i = -1
-    while (++i < length) {
+    while (++i < length)  {
```
[the remark this answers is not in the document]
Comment (uroybd) [reply index.js]: I'm just curious, what's the benefit of using while instead of for or map in this case?

[the remark this answers is not in the document]
Comment (yuanqing) [reply index.js]: No benefit other than perhaps that it's concise. But really it’s just a convention in this code base for doing loops

**Comment by yuanqing** [comment]
Thanks @uroybd!

**Comment by uroybd** [comment]
You are welcome! @yuanqing

File: index.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (uroybd) [reply index.js]: Ah! I find for loop with inline var because garbage collector will automatically collect that variable at the end. Therefore, no residue. Nowadays I use map instead of for if I don't have to break it in middle somewhere which is more efficient. and gives me more control over data. :)

**Comment by yuanqing** [comment]
Just cut a `2.2.0`
